Feature/facets

data-envoy opened the pull request:
This PR implements facets and filtering on data within a single table.

- I also thought very hard about the best way to configure. I've mirrored Algolina search API and setting API. It should work with zero-conf but has the option of configuring before production to improve perf and security. It's typed with autofill/autocomplete for the end developer. An explination is in the readme.
- A new demo dataset, that's more accessible with 20K records.
- Cleaner imports at the top of index.ts

In client.ts I renamed... queries: Query[] -> requests: Request[] . To try and avoid ambiguity, because there is a `query` property. But, I think `requests` is a poor choice, as it's ambigous with http requests.

BTW. This includes the highlight PR, in case you want to ignore that.

Fixes https://github.com/dekimir/postgres-searchbox/issues/13

data-envoy committed:
Facets. Partially complete. Where clause from request.

data-envoy committed:
delete coverage

data-envoy committed:
Ignore coverage dir

data-envoy committed:
Facets bug fix: facet request params are already parsed in index

data-envoy committed:
Missed updating test from last commit.

data-envoy committed:
Merge branch 'dekimir-main' into feature/facets

data-envoy committed:
Merge branch 'feature/typescript'

# Conflicts:
#	.gitignore
#	examples/with-nextjs/next.config.js
#	examples/with-nextjs/pages/api/search.ts
#	package/package.json

data-envoy committed:
Merge branch 'feature/typescript'

data-envoy committed:
Progress checkin - unfinished

data-envoy committed:
examples, add store

data-envoy committed:
Install dts-bundle-generator for inlining dependency types
Add create-store script
Update readme
Add setting for server side config and search options for client config
Faces and filters in unique files
Fix typos

data-envoy committed:
Update readme with terminology and config.

data-envoy committed:
Update examples and readme typos

data-envoy committed:
typos

data-envoy committed:
Delete examples readme

data-envoy committed:
Merge branch 'dekimir-main' into feature/facets

# Conflicts:
#	examples/with-nextjs/yarn.lock
#	package/README.md

data-envoy committed:
Update readme, re highlight config.

data-envoy committed:
Return facets_stats on request response, allows for instantsearch RangeInput component

dekimir commented:
Thank you.  Let's work on the highlight PR first, to get that merged soon.

data-envoy committed:
FacetSearch with validation. Example updated.

data-envoy committed:
Write test for facet search

data-envoy committed:
Simplify the lib/filters function

dekimir commented:
#9 has now landed in `main`, so please rebase on that, for easier reviewing.

data-envoy committed:
Merge upstream and remove unfinished extendDatabase code

data-envoy commented:
Great, I'll reply to the review later :). This has been rebased.

data-envoy committed:
Facet search: Fix case on highlighting.

dekimir commented on examples/with-nextjs/pages/api/search.ts:
```diff
@@ -1,24 +1,71 @@
 import { getSearchHandler } from 'postgres-searchbox';
 // During postgres-searchbox development this can be:
-// import { searchHandler } from '../../../../package/build';
+// import { getSearchHandler } from '../../../../package/build/index.js';
```
Interesting. Did importing from just `package/build` ever work? If so, what's changed?

data-envoy replied:
Without .js works now. I think I just needed to restart the TS server. I've done that for all in example/with-nextjs to keep them consistent throughout the example.

dekimir commented on package/README.md:
```diff
@@ -133,28 +134,128 @@ By default Postgres sorts asc and returns null values first. So they can be left
+but are only a subset of Algolia. They can be set with type-safety and autofill `ctrl + space` in VSCode.
+
+If your searchHandler should handle multiple indexes, instead of passing one config object you can pass in an
+array of configs like this. Make sure to set the indexName property for each config.
```
... and make sure to run the create-index script for each such name.

data-envoy replied:
Good point.

dekimir commented on package/README.md:
```diff
@@ -224,12 +330,14 @@ postgres-searchbox installed. See the 3 files:
+- You can see the movies page at http://locaalhost:3000/movies
+- You can see the store page at http://locaalhost:3000/store
+
+NextJS can import the `package/build/*.js` files, to keep them up to date run `yarn dev` from a 2nd terminal.
```
I don't understand why the second `yarn dev` is necessary. If you uncomment the local-dir imports (`from '.../package/build'`) in `pages/`, NextJS will automatically rebuild on any change in the package.

data-envoy replied:
This is related to https://github.com/dekimir/postgres-searchbox/pull/12#discussion_r1180483442 . I don't think I know the best way to do this.

At the momement the yarn dev is needed in the package dir because nextjs is importing from the build directory. Without it, the build directory files don't get updated. So nextjs doesn't know when to rebuild.

dekimir replied:
Ah, that makes sense. Then this is fine. I'll rephrase it later to make it a bit clearer, but what it says is true.

dekimir commented on examples/with-nextjs/pages/api/search.ts:
```diff
@@ -1,24 +1,71 @@
+      ],
+      attributesForFaceting: [
+        'brand',
+        'hierarchicalCategorieslvl0',
```
Hm, does this work the same as in Algolia? It's not trivial to hit the correct level of hierarchy to expose in the most general case. (Eg: you have a country/province/city hierarchy and your search results all happen to be from the same province.)

data-envoy replied:
I'm not 100% sure I understand your question. I think it's comparable to Algolia. I've compared their [demo store](https://instantsearchjs.netlify.app/examples/react/e-commerce/search/Appliances/?query=Dehumidifier) to this one.

With an example of Air Conditioners being in the middle of the hierachy: `Appliances > Air Conditioners > Portable Air Conditioners`. And searching for 'air conditioner' 

The behaviour looks similar, but not identical, in part to fuzzy matching and other Algolia features. Nothing stands out as looking wrong.

Related, I've added to the readme on how to store the hierachal data.

dekimir commented on examples/with-nextjs/pages/movies.tsx:
```diff
@@ -24,19 +24,30 @@ import {
+import type { SearchOptions } from 'postgres-searchbox/client.types';
 // During postgres-searchbox development this can be:
 // import { make_client } from '../../../package/build/client';
+// import type { SearchOptions } from '../../../package/build/client.types';
```
So why don't you need `.js` at the end of these two lines, like you did in `api/search.ts`?

data-envoy replied:
This is something that's been difficult for me to understand. I'll try and work out a better way.

dekimir commented on examples/with-nextjs/pages/store.tsx:
```diff
@@ -0,0 +1,170 @@
+  };
+
+  return (
+    <>
```
Again, I don't know what this `<>` does? Did you add it, or is it just generated boilerplate?

Edit: Ha!
https://github.com/vercel/next.js/blob/c94122dc2c417d5be11b0bfa50c0957e0e2ae91a/packages/create-next-app/templates/default/ts/pages/index.tsx#L10

data-envoy replied:
I guess you figured it out... It's so a react function can return 2 components. :) Well, it wraps them into a single one. In this specific case, so `<Head>` and `<main>` are returned.

dekimir replied:
Here's official doc on this, for future reference:
https://react.dev/reference/react/Fragment

dekimir commented on package/package.json:
```diff
@@ -56,15 +60,16 @@
+    "rimraf": "^5.0.0",
     "ts-node": "^10.9.1",
-    "typescript": "^5.0.4",
-    "zod": "^3.21.4"
```
No more `zod`?? But we use `zod`! :)

data-envoy replied:
This is just de-duplicaing it as it was in both devDeps. and deps.

dekimir replied:
OK.

dekimir commented on package/scripts/create-index.ts:
```diff
@@ -28,6 +28,9 @@ export async function createColumnAndIndex({
   // Get column names
   const columnNames = await getTextColumnsFromTable({ tableName });
+  if (!columnNames?.length) {
+    throw 'Found no text columns in createColumnAndIndex > getTextColumnsFromTable';
```
Replace `createColumnAndIndex > getTextColumnsFromTable` with `${tableName}`. It's more useful to the end user and more future-proof.

data-envoy replied:
I agree, fixed :)

dekimir commented on package/scripts/create-index.ts:
```diff
@@ -78,7 +81,8 @@ export async function dropColumnAndIndex({ tableName }: { tableName: string }) {
  * Run the script when called from package.json
- * Self init functions because they are async.
+ * Self init functions because top-level await is
+ * un-supported in node < v14.8.0.
```
First off, we require node > v18, because we use global fetch, right? The whole package won't work with older Node versions, so this seems unnecessary.

And if the comment stays nevertheless, please move it down to line 93, closer to the functions it describes. Right now it looks like it applies to `errorString`, which is confusing.

dekimir replied:
To be clear: I'm not opposed to supporting older Node versions, it just seems to me we currently don't.

data-envoy replied:
I tried to get rid of these but the real issue is with getting the top-level await to work with @jest/swc. I've moved and updated the comments

dekimir commented on package/scripts/create-movies.ts:
```diff
@@ -113,21 +113,10 @@ export async function importData({ tableName }: { tableName: string }) {
  * Run the script when called from package.json
- * Self init functions because they are async.
+ * Self init functions because top-level await is
+ * un-supported in node < v14.8.0.
```
Ditto.

dekimir commented on package/scripts/create-store.ts:
```diff
@@ -0,0 +1,68 @@
+ * Run the script when called from package.json
+ * Self init functions because top-level await is
+ * un-supported in node < v14.8.0.
+ */
```
Here, too.

dekimir commented on package/src/index.ts:
```diff
@@ -1,128 +1,412 @@
+    // Don't return the error stack to the browser
+    const browserError =
+      errors[0].name === 'ZodError'
+        ? 'Request contained invalid payload'
```
It wouldn't hurt to put in the response what specifically is invalid. That would help requesters who don't have access to the console, for whatever reason.

data-envoy replied:
I've updated the code so zod errors are passed to the client

dekimir commented on package/src/lib/filters.ts:
```diff
@@ -0,0 +1,258 @@
+ * Both the <DynamicWidgets> and <RefinementList> components make request with this format
+ */
+
+export const getFilters = async ({
```
Why is this async?

data-envoy replied:
Fixed :)

dekimir commented on package/src/index.ts:
```diff
@@ -1,128 +1,412 @@
-    ), query AS (
-      SELECT 
+      WHERE
+        ( ( %I @@ websearch_to_tsquery(%L) AND %L <> '' )  OR %L = '' )
```
Why are the extra AND and OR necessary now? Can you add a comment?

data-envoy replied:
This is me being an sql rookie. It's so we get *all* results on an empty query. 
It should be: `( %I @@ websearch_to_tsquery(%L)  OR %L = '' )` I'll update

dekimir replied:
Doesn't an empty query @@-match all rows anyway?

data-envoy replied:
No, both are always false:
`postgres_searchbox_v1_doc @@ '', `
`postgres_searchbox_v1_doc @@ websearch_to_tsquery('')`

dekimir replied:
Ah, OK.

data-envoy committed:
index.test.ts: Move init db code to beforeEach
Add pagination test for offset including default length

dekimir commented on package/src/lib/facets.ts:
```diff
@@ -0,0 +1,133 @@
+            '{}'::json
+          ) as details
+            FROM (
+              SELECT %s, count(*) AS cnt
```
We should make this alias specific to postgres-searchbox. What if the facet's name also happens to be `cnt`?

This applies to other such situations, too.

data-envoy replied:
This is a great shout, I found another place to update in lib/highlight.ts - I think that's all.

data-envoy committed:
Remove .only from 2 tests

dekimir commented on package/test/integration.test.ts:
```diff
@@ -56,11 +61,11 @@ describe('integration', () => {
-    const dropSql = format('DROP TABLE IF EXISTS %I', tableName);
-    await client.query(dropSql);
+    // const dropSql = format('DROP TABLE IF EXISTS %I', tableName);
+    // await client.query(dropSql);
```
Let's delete these lines instead of commenting them out.

data-envoy replied:
oops, this was commented by mistake, probably during debugging. I will uncomment & keep.

dekimir commented on package/test/integration.test.ts:
```diff
@@ -74,7 +79,7 @@ describe('integration', () => {
       {
         indexName: tableName,
-        params: { query: 'affordable keyboard' },
+        params: { query: 'bespoke keyboard' },
```
Why change this?

data-envoy replied:
This was related to finding keywords that were in the name and description. Adding to createRandomProduct changed the database result. I've updated that script so changes like this shouldn't be needed in the future.

dekimir commented on package/test/__snapshots__/index.test.ts.snap:
```diff
@@ -1,26 +1,75 @@
-        "fullyHighlighted": false,
-        "matchLevel": "full",
-        "matchedWords": [],
-        "value": "__ais-highlight__Football__/ais-highlight__ __ais-highlight__awesome__/ais-highlight__ Is Good For Training And Recreational Purposes",
```
Why has this result changed? The `football` query is still in the test.

data-envoy replied:
Similar to some of the other keyword changes.. 
> Adding to createRandomProduct changed the database result. Shouldn't happen again.

dekimir replied:
That only explains the *addition* (Carbonite gloves) to the snapshot. It doesn't explain the *deletion* (football) in it. I still don't understand why this test searches for "football" but doesn't expect "football" in the result?

Also, I find it naive to think that this won't happen again. Adding more features tomorrow will likely require more fiddling with `faker` to make those features testable.

data-envoy replied:
The reason the football query was dropped is because I wanted to test a query that highligted words in both the name and description. And with the updated faker seed, football wasn't meeting that requirement anymore.

With the revised code in `createRandomProduct` additional properties can be generated as long as they are added to the end of the function. These won't have the knock on effect of offsetting the seed. I've added a comment. 

I can't really think of a better way to have reproducable and semi-realistic test data. I'm open to suggestions.

dekimir replied:
> The reason the football query was dropped

But the football query wasn't dropped, that's my whole point.  It's still right there in `index.test.ts`: https://github.com/data-envoy/postgres-searchbox/blob/5065a415993f3cb904c02c3b4ced9d01e9ce505b/package/test/index.test.ts#L309

> With the revised code in `createRandomProduct` additional properties can be generated as long as they are added to the end of the function. These won't have the knock on effect of offsetting the seed. I've added a comment.

Great.
 
> I can't really think of a better way to have reproducable and semi-realistic test data. I'm open to suggestions.

This is more than good enough for now.  One day I'm envisioning explicitly populating the data (we can run `faker` once and record the results), so we can see exactly what's in the corpus and how each query matches it.

data-envoy replied:
Ah I think I know what happened. For highlighting I changed the query to 'bespoke ball' to get highlighting in both name and desc.

I left it as football in the `should handle returning columns` test, because that test the search query is only testing that the right columns are being returned.

I appreciate an explination in the description would have been more helpful, I'll keep that in mind for future work!

> One day I'm envisioning explicitly populating the data (we can run faker once and record the results)

That sounds like a good plan :D

dekimir replied:
OK.

dekimir reviewed (changes requested):
Looks good. I just want to understand why the test snapshots changed.

data-envoy committed:
|| change to ??

data-envoy committed:
lib/columns.ts to return db.formatted: string []

data-envoy committed:
Remove references to extendedAttributes.
Remove expect(consoleSpy).toHaveBeenCalled();
Simplify sql line websearch_to_tsquery line

data-envoy commented:
Adding to the list here

- Simplify complicated mock data logic https://github.com/dekimir/postgres-searchbox/pull/9#discussion_r1180399859
- Transparent zod errors to client https://github.com/dekimir/postgres-searchbox/pull/9#discussion_r1180409384
- Add yarn build to a test in scripts.test.ts . Else al tests could pass even with the project having Typescript errors.

data-envoy committed:
Revert integration.test.ts commented code

data-envoy commented:
The snapshots changed because the when I added a new faker call like`faker.company.name()` to `createRandomProduct`. It knocked all of the subsequent faker calls out of sync.

I think the seed number increases by one on every faker.x() call. I will try and address this, but the snapshots will change one more final time.

data-envoy committed:
Simplify and improve faker generated items to be more consistent.
Update tests and snapshots accordingly.

data-envoy committed:
Remove unnecessary use of async on functions.
Better error message

data-envoy committed:
Update derived column names so as not to collide with user's columns.
Update readme re multiple configs.
https://github.com/dekimir/postgres-searchbox/pull/12#discussion_r1180358907

dekimir commented on package/scripts/mock-data.ts:
```diff
@@ -58,16 +51,13 @@ export async function initTestDatabase({
   await client.query(sql);
-  // Set deed so that values are always the same
-  if (fakerSeed) {
-    faker.seed(123);
```
You should add "fixes #13" to this PR's description. ;)

data-envoy committed:
Update readme to cover HierarchicalMenu

data-envoy committed:
Fix readme typos

data-envoy committed:
Zod errors are transparently passed to client
DB and server errors logged but not passed directly to client
Tests updated accordingly
Improve index validation with better validation and using zod built in validation insteadd of some custom refines

data-envoy committed:
Add yarn build to CLI test suite

data-envoy committed:
Update top-level await comments

data-envoy committed:
Fix typos

data-envoy commented:
@dekimir thanks for the review. I think everything has been addressed here.

data-envoy committed:
in example. Remove unnecessary .js extension on imports
